src/pepr/operator/crd/validators/package-validator.ts

```diff
@@ -92,6 +92,32 @@ export async function validator(req: PeprValidateRequest<UDSPackage>) {
         `The client ID "${client.clientId}" uses an invalid secret name ${client.secretName}`,
       );
     }
+    // If standardFlowEnabled is undefined (defaults to `true`) or explicitly true and there are no redirectUris set, deny the req
+    if (
+      (client.standardFlowEnabled === undefined || client.standardFlowEnabled) &&
+      !client.redirectUris
+    ) {
+      return req.Deny(
+        `The client ID "${client.clientId}" must specify redirectUris if standardFlowEnabled is turned on`,
+      );
+    }
+    // If this is a public client ensure that it only sets itself up as an OAuth Device Flow client
+    if (
+      client.publicClient &&
+      (client.standardFlowEnabled === undefined ||
+        client.standardFlowEnabled ||
+        client.secret ||
+        client.secretName ||
+        client.secretTemplate ||
+        client.enableAuthserviceSelector ||
+        client.protocol === "saml" ||
+        !client.attributes ||
+        client.attributes["oauth2.device.authorization.grant.enabled"] != "true")
```

Comment:
I think this should work to block misconfiguring a client to be public but still attempt use a secret. That scenario is actually an error.

It's really a separate _policy_ thing though to block all public clients with the standard flow though. All other policy things in UDS usually have a way to say "please bypass this policy please". I guess we could check an annotation on the Package CR? Seems weird though. There are other plausible use cases for public clients, but we UDS doesn't have them currently, I'm unsure what kind of weird Delivery use case might need this.

For now, here's some cleanup around the tests below. Notably, in JS the empty string is falsy, I think we need `!== undefined` to ensure things are actually entirely unset (setting them to the empty string should be disallowed too).
```suggestion
    if (
      client.publicClient &&
      (client.standardFlowEnabled !== false ||
        client.secret !== undefined ||
        client.secretName !== undefined ||
        client.secretTemplate !== undefined ||
        client.enableAuthserviceSelector !== undefined ||
        client.protocol === "saml" ||
        client.attributes?.["oauth2.device.authorization.grant.enabled"] !== "true")
```

Reply:
applied your suggestion - we could cross that bridge when we get there and just not allow it for now until someone needs it - I can't really think of anything where you'd want a public client for a normal web app - the only time you might want the standard auth flow and a public client would be the default flow for cosign / witness but in all of the use cases I can think of you can use the device flow instead - it does slightly hurt UX since you have to have step or add a command flag depending on the client but I think that is largely ok (and we could wrap things for them in our own stuff too))

Reply:
Are there unit tests in UDS operator? Do we usually do them for things like this?
Can we test with each bad field set to both the "wrong" value (true/false) and also to `""`?
Or just do manual testing I guess.

I just don't actually trust myself (or the insanity that is JS) to write complex `if` statements. I want to make sure we've actually tested this.

Reply:
I can add some in there - would likely be similar to the exempt validator

Reply:
Added tests covering the package validator